src/dependencies/dss-direct-deposit/D3MInit.sol

```diff
@@ -0,0 +1,326 @@
+    function implementation() external view returns (address);
+}
+
+interface D3MOracleLike {
+    function vat() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function file(bytes32, address) external;
+}
+
+interface D3MHubLike {
+    function vat() external view returns (address);
+    function daiJoin() external view returns (address);
+    function file(bytes32, address) external;
+    function file(bytes32, bytes32, address) external;
+    function file(bytes32, bytes32, uint256) external;
+}
+
+interface D3MMomLike {
+    function setAuthority(address) external;
+}
+
+struct D3MCommonConfig {
+    address hub;
+    address mom;
+    bytes32 ilk;
+    bool existingIlk;
+    uint256 maxLine;
+    uint256 gap;
+    uint256 ttl;
+    uint256 tau;
+}
+
+struct D3MAavePoolConfig {
+    address king;
+    address adai;
+    address stableDebt;
+    address variableDebt;
+}
+
+struct D3MAaveRateTargetPlanConfig {
+    uint256 bar;
+    address adai;
+    address stableDebt;
+    address variableDebt;
+    address tack;
+    uint256 adaiRevision;
+}
+
+struct D3MAaveBufferPlanConfig {
+    uint256 buffer;
+    address adai;
+}
+
+struct D3MCompoundPoolConfig {
+    address king;
+    address cdai;
+    address comptroller;
+    address comp;
+}
+
+struct D3MCompoundRateTargetPlanConfig {
+    uint256 barb;
+    address cdai;
+    address tack;
+    address delegate;
+}
+
+// Init a D3M instance
+library D3MInit {
+
+    function initCore(
+        DssInstance memory dss,
+        D3MCoreInstance memory d3mCore
+    ) internal {
+        D3MHubLike hub = D3MHubLike(d3mCore.hub);
+        D3MMomLike mom = D3MMomLike(d3mCore.mom);
+
+        // Sanity checks
+        require(hub.vat() == address(dss.vat), "Hub vat mismatch");
+        require(hub.daiJoin() == address(dss.daiJoin), "Hub daiJoin mismatch");
+
+        hub.file("vow", address(dss.vow));
+        hub.file("end", address(dss.end));
+
+        mom.setAuthority(dss.chainlog.getAddress("MCD_ADM"));
+
+        dss.vat.rely(address(hub));
+
+        dss.chainlog.setAddress("DIRECT_HUB", address(hub));
+        dss.chainlog.setAddress("DIRECT_MOM", address(mom));
+    }
+
+    function initCommon(
+        DssInstance memory dss,
+        D3MInstance memory d3m,
+        D3MCommonConfig memory cfg
+    ) internal {
+        bytes32 ilk = cfg.ilk;
+        D3MHubLike hub = D3MHubLike(cfg.hub);
+        D3MOracleLike oracle = D3MOracleLike(d3m.oracle);
+
+        // Sanity checks
+        require(oracle.vat() == address(dss.vat), "Oracle vat mismatch");
+        require(oracle.ilk() == ilk, "Oracle ilk mismatch");
+
+        WardsAbstract(d3m.plan).rely(cfg.mom);
+
+        hub.file(ilk, "pool", d3m.pool);
+        hub.file(ilk, "plan", d3m.plan);
+        hub.file(ilk, "tau", cfg.tau);
+
+        oracle.file("hub", address(hub));
+
+        dss.spotter.file(ilk, "pip", address(oracle));
+        dss.spotter.file(ilk, "mat", 10 ** 27);
```

Comment:
I'd avoid magical values here.

At least document it like:
```
// 100% liquidation ration
```

Maybe extract this into a constant/memory variable like `ONE_HUNDRED_PERCENT_RAY = 1 * 10 ** 27`.

Reply:
Same here, this is coming from the D3M repo. It should potentially first be changed there via a PR, then use the updated version on spells. https://github.com/makerdao/dss-direct-deposit/tree/master/src/deploy

Reply:
Can update in d3m repo. Can we go with this as-is, and I'll add a PR there is another small fix I want to do as well. PS this type of stuff (inside the init library) should be reviewed/audited inside the repo and well in advance of the spell imo.

Reply:
Added to PR: https://github.com/makerdao/dss-direct-deposit/pull/109

LMK if you want to update to this before the spell goes out.